File: web-canvas/dom-api/src/jsMain/kotlin/pl/karol202/uranium/webcanvas/dom/window/NativeWindow.kt
package pl.karol202.uranium.webcanvas.dom.window

import pl.karol202.uranium.webcanvas.dom.input.NativeKeyboardEvent
import kotlin.browser.window

actual object NativeWindow
{
	actual fun now() = window.performance.now()

	actual fun setInterval(interval: Int, handler: () -> Unit) = window.setInterval(handler, interval)

	actual fun clearInterval(handle: Int) = window.clearInterval(handle)

	actual fun setOnKeyDownListener(listener: ((NativeKeyboardEvent) -> Unit)?)
	{
		window.onkeydown = { listener?.invoke(NativeKeyboardEvent(it)) }
	}

	actual fun setOnKeyPressListener(listener: ((NativeKeyboardEvent) -> Unit)?)
	{
		window.onkeypress = { listener?.invoke(NativeKeyboardEvent(it)) }
	}

	actual fun setOnKeyUpListener(listener: ((NativeKeyboardEvent) -> Unit)?)
	{
		window.onkeyup = { listener?.invoke(NativeKeyboardEvent(it)) }
	}
}

File: web-canvas/dom-api/src/wasm32Main/js/nativeWindow.js
konan.libraries.push({
    uranium_NativeWindow_now: () =>
        window.performance.now(),
    uranium_NativeWindow_setInterval: (timeout, handler) =>
        window.setInterval(handler, timeout),
    uranium_NativeWindow_clearInterval: (handle) =>
        window.clearInterval(handle),
    uranium_NativeWindow_setOnKeyDownListener: (handler) =>
        window.onkeydown = handler,
    uranium_NativeWindow_setOnKeyPressListener: (handler) =>
        window.onkeypress = handler,
    uranium_NativeWindow_setOnKeyUpListener: (handler) =>
        window.onkeyup = handler
});
